vars/withSelenium.groovy

```diff
@@ -0,0 +1,176 @@
+ *      default network.
+ * @param closure the body
+ */
+void call(Map config = [:], String seleniumNetwork, Closure closure) {
+
+    def defaultConfig = [
+            seleniumHubImage  : 'selenium/hub',
+            seleniumVersion   : "3.141.59-zinc",
+            workerImageFF     : "selenium/node-firefox",
+            workerImageChrome : "selenium/node-chrome",
+            firefoxWorkerCount: 0,
+            chromeWorkerCount : 0,
+            debugSelenium     : false
+    ]
+
+    def networkExists = checkNetwork(seleniumNetwork)
+    if(!networkExists) {
+        throw new ConfigurationException("the given network '${seleniumNetwork}' does not exist but is mandatory when using selenium grid")
+    }
+
+    // Merge default config with the one passed as parameter
+    config = defaultConfig << config
+    if (config.firefoxWorkerCount == 0 && config.chromeWorkerCount == 0) {
+        throw new ConfigurationException("Cannot start selenium test. Please configure at least one workerCount for the desired browser.")
+    }
+
+    def uid = findUid()
+    def gid = findGid()
+
+    def networkParameter = ""
+    if (seleniumNetwork != null && !seleniumNetwork.isEmpty()) {
+        networkParameter = "--network ${seleniumNetwork}"
+    }
+
+    gridDebugParameter = ""
+    if (gridDebugParameter != null && !gridDebugParameter.isEmpty()) {
+        gridDebugParameter = "-e GRID_DEBUG=true"
+    }
+
+    String hubName = generateJobName() + "-seleniumhub"
+
+    // explicitly pull the image into the registry. The documentation is not fully clear but it seems that pull()
+    // will persist the image in the registry better than an docker.image(...).runWith()
+    def seleniumHubImage = docker.image("${config.seleniumHubImage}:${config.seleniumVersion}")
+    seleniumHubImage.pull()
+    // Run with Jenkins user, so the files created in the workspace by selenium can be deleted later
+    // Otherwise that would be root, and you know how hard it is to get rid of root-owned files.
+    def dockerArgs = "-u ${uid}:${gid} ${networkParameter} ${gridDebugParameter} -p 4444:4444 --name ${hubName}"
+    seleniumHubImage.withRun(dockerArgs) { hubContainer ->
+        String seleniumIp = findContainerIp(hubContainer)
+
+        def firefoxContainers = runWorkerNodes("${config.workerImageFF}:${config.seleniumVersion}", networkParameter, gridDebugParameter, seleniumIp, config.firefoxWorkerCount)
+        def chromeContainers = runWorkerNodes("${config.workerImageChrome}:${config.seleniumVersion}", networkParameter, gridDebugParameter, seleniumIp, config.chromeWorkerCount)
+
+        try {
+            waitForSeleniumToGetReady(seleniumIp)
+
+            closure.call(hubContainer, seleniumIp, uid, gid)
+        } finally {
+            stopSeleniumSession(firefoxContainers, chromeContainers)
+        }
+    }
+}
+
+String generateJobName() {
```

Comment:
A large part of them methods here and downward are utility methods with no direct connection to running selenium, and I swear I saw them somewhere else too: :thinking: Could they be split off and shared among the library parts? They surely are used there too, aren't they?

Reply:
I tried to extract this function but I did not get it to work. Maybe the `import` and `evaluate` mechanisms work diffently for jenkins.